### PR TITLE
refactor(api): Allow tip tracking for 8/96 channel pipette

### PR DESCRIPTION
**Pull request by TamarZanzouri**
# Overview

closes https://opentrons.atlassian.net/browse/RLIQ-267.
Add tip tacking support for 8/96 channels via protocol engine tip state.

# Changelog

- Added `channels_by_pipette_id` to `TipState` in order to know how many channels are associated to the given pipette id. 
- Changed `column_heads_by_labware_id` to `column_by_labware_id` and storing wells per column.
- Added support of `AddPipetteConfigAction` in handle_action to initialize `channels_by_pipette_id`.
- changed `get_next_tip` to work with columns instead of column heads. 

# Review requests

1. should I create a new dispatched action for the pipette channels and remove the channels completely from `StaticPipetteConfig`? it is currently duplicated with `PipetteState`.
2. am I missing tests? 
3. does logic makes sense?

# Risk assessment

Medium. TipState now hashes columns but `get_next_tip` still checks against column heads.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 1]
```diff
@@ -17,13 +22,18 @@ class TipRackWellState(Enum):
 
 TipRackStateByWellName = Dict[str, TipRackWellState]
 
+_96_CHANNEL_PIPETTE = 96
+_8_CHANNEL_PIPETTE = 8
+
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]:
```suggestion
```

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 2]
```diff
@@ -73,6 +83,36 @@ def handle_action(self, action: Action) -> None:
                     well_name
                 ] = TipRackWellState.CLEAN
 
+        elif isinstance(action, AddPipetteConfigAction):
+            self._state.channels_by_pipette_id[
+                action.pipette_id
+            ] = action.static_config.channels
+
+    def _update_used_tips(
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]:
```suggestion
    def _set_used_tips(
```

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 3]
```diff
@@ -93,7 +133,8 @@ def get_next_tip(
         """Get the next available clean tip."""
         seen_start = starting_tip_name is None
         wells = self._state.tips_by_labware_id[labware_id]
-        column_heads = self._state.column_heads_by_labware_id[labware_id]
+        columns = self._state.column_by_labware_id[labware_id]
+        column_heads = [column[0] for column in columns]
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]: Should we make sure as part of this PR the for 8 channels the whole column is available?

File: api/tests/opentrons/protocol_engine/state/test_tip_state.py [hunk 4]
```diff
@@ -1,14 +1,18 @@
...
 from opentrons_shared_data.labware.labware_definition import LabwareDefinition
 
 from opentrons.protocol_engine import actions, commands
+from opentrons.protocol_engine.types import StaticPipetteConfig
 from opentrons.protocol_engine.state.tips import TipStore, TipView
 
 
 @pytest.fixture
-def subject() -> TipStore:
+def subject(decoy: Decoy) -> TipStore:
```
Comment (TamarZanzouri) [inline api/tests/opentrons/protocol_engine/state/test_tip_state.py]:
```suggestion
def subject() -> TipStore:
```

File: api/tests/opentrons/protocol_engine/state/test_tip_state.py [hunk 5]
```diff
@@ -177,3 +206,32 @@ def test_reset_tips(
     )
 
     assert result == "A1"
+
+
+@pytest.mark.parametrize("input_channels, next_tip_result", [(96, None), (8, "A2")])
```
Comment (TamarZanzouri) [inline api/tests/opentrons/protocol_engine/state/test_tip_state.py]: Should I add a separate test for the dispatched config action?

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 6]
```diff
@@ -17,13 +22,18 @@ class TipRackWellState(Enum):
 
 TipRackStateByWellName = Dict[str, TipRackWellState]
 
+_96_CHANNEL_PIPETTE = 96
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]: is this redundant?

File: api/src/opentrons/protocol_api/core/engine/instrument.py [hunk 7]
```diff
@@ -363,7 +363,7 @@ def get_hardware_state(self) -> PipetteDict:
         return self._sync_hardware_api.get_attached_instrument(self.get_mount())  # type: ignore[no-any-return]
 
     def get_channels(self) -> int:
-        return self._engine_client.state.pipettes.get_channels(self._pipette_id)
+        return self._engine_client.state.tips.get_channels(self._pipette_id)
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/core/engine/instrument.py]: curious why `get_channels` is on the `tips` state rather than the `pipettes` state.

File: api/src/opentrons/protocol_engine/actions/actions.py [hunk 8]
```diff
@@ -175,7 +175,10 @@ class AddPipetteConfigAction:
     """Adds a pipette's static config to the state store."""
 
     pipette_id: str
-    static_config: StaticPipetteConfig
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/actions/actions.py:178]: I think we want this to still be a `static_config`, right?

File: api/src/opentrons/protocol_engine/state/tips.py (hunk 6 above)
Comment (Laura-Danielle) [reply api/src/opentrons/protocol_engine/state/tips.py]: I think the max number of channels should be a property on pipette state that we get from `static configs`. Eventually when we do support partial tip pick up, there will be a stateful pipette property like `working channels`.

**Review by Laura-Danielle — COMMENTED** [review]
couple initial comments.

File: api/src/opentrons/protocol_api/core/engine/instrument.py (hunk 7 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/core/engine/instrument.py]: Because at the moment the channel count is only used in the engine itself for updating tip state. It'll probably continue to shift around as we add more functionality

File: api/src/opentrons/protocol_engine/actions/actions.py (hunk 8 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/actions/actions.py:178]: This change is just removing an intermediate object in favor of putting the payload directly on the `AddPipetteConfigAction` dataclass

File: api/src/opentrons/protocol_engine/state/tips.py (hunk 6 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/tips.py]: I think these constants are fine for now, since they're used internally to compare to `pipette_channels` from the config. We can change as needed when we do partial pickup.

I don't _think_ we'll need any additional information from the HW API for this, especially if partial pickups always continue to use the rear-left channel

File: api/tests/opentrons/protocol_engine/state/test_tip_state.py (hunk 5 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_tip_state.py]: Yes, there needs to be a separate test that dispatches an `AddPipetteConfigAction` to the store and checks that `get_channels` returns the added config

File: api/src/opentrons/protocol_engine/state/tips.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/tips.py]: Yes, it looks like we need to make two additions to this PR before it's ready

- If `next_tip` is requested for an 8-channel, find the next available column (checking if all wells in the column have a tip)
- If requested for a 96 channel, check that _all_ tips in the tip rack are free

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 9]
```diff
@@ -73,6 +84,32 @@ def handle_action(self, action: Action) -> None:
                     well_name
                 ] = TipRackWellState.CLEAN
 
+        elif isinstance(action, AddPipetteConfigAction):
+            self._state.channels_by_pipette_id[action.pipette_id] = action.channels
+
+    def _set_used_tips(self, pipette_id: str, well_name: str, labware_id: str) -> None:
+        pipette_channels = self._state.channels_by_pipette_id.get(pipette_id)
+        if pipette_channels == _96_CHANNEL_PIPETTE:
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/tips.py]: As mentioned in my other comment, I think this is fine, but we could compare against the column length and the total well count of the labware itself rather than encoding `8` and `96` in constants.

For example:
- If channel count == column length, use whole column
- Elif channel count == total well count, use whole labware

**Review by mcous — COMMENTED** [review]
Looks like we're still missing some required functionality, but we're getting there!

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 10]
```diff
@@ -73,6 +81,28 @@ def handle_action(self, action: Action) -> None:
...
+    def _set_used_tips(self, pipette_id: str, well_name: str, labware_id: str) -> None:
+        pipette_channels = self._state.channels_by_pipette_id.get(pipette_id)
+        columns = self._state.column_by_labware_id[labware_id]
+        tips_in_labware = self._state.tips_by_labware_id[labware_id]
+
+        if pipette_channels == len(tips_in_labware.keys()):
+            for well_name in tips_in_labware.keys():
+                tips_in_labware[well_name] = TipRackWellState.USED
+
+        elif pipette_channels == len(columns[0]):
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]:
```suggestion
        elif columns and pipette_channels == len(columns[0]):
```

File: api/src/opentrons/protocol_api/core/engine/labware.py [hunk 11]
```diff
@@ -120,7 +120,7 @@ def get_next_tip(
     ) -> Optional[str]:
         return self._engine_client.state.tips.get_next_tip(
             labware_id=self._labware_id,
-            use_column=num_tips != 1,
+            tip_amount=num_tips,
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/engine/labware.py]: If we're going to pass this argument through, lets keep it named the same thing so our brains can do one less thing

```suggestion
            num_tips=num_tips,
```

File: api/src/opentrons/protocol_engine/commands/load_labware.py [hunk 12]
```diff
@@ -63,7 +63,8 @@ class LoadLabwareResult(BaseModel):
     )
     offsetId: Optional[str] = Field(
         # Default `None` instead of `...` so this field shows up as non-required in
-        # OpenAPI. The server is allowed to omit it or make it null.
+        # OpenAPI. The server is
+        # allowed to omit it or make it null.
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/load_labware.py]: Unrelated change?

File: api/src/opentrons/protocol_engine/state/pipettes.py [hunk 13]
```diff
@@ -87,7 +87,11 @@ def handle_action(self, action: Action) -> None:
         elif isinstance(action, SetPipetteMovementSpeedAction):
             self._state.movement_speed_by_id[action.pipette_id] = action.speed
         elif isinstance(action, AddPipetteConfigAction):
-            self._state.static_config_by_id[action.pipette_id] = action.static_config
+            self._state.static_config_by_id[action.pipette_id] = StaticPipetteConfig(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/pipettes.py:99]: I would consider moving the `StaticPipetteConfig` class here, if this is the only place it's used

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 14]
```diff
@@ -48,22 +59,19 @@ def handle_action(self, action: Action) -> None:
                 for column in definition.ordering
                 for well_name in column
             }
-            self._state.column_heads_by_labware_id[labware_id] = set(
-                column[0] for column in definition.ordering
+            self._state.column_by_labware_id[labware_id] = list(
+                column for column in definition.ordering
             )
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/tips.py]: Unnecessary call to `list` constructor, prefer `[...]` literals. Also prefer name `columns` to match existing conventions


```suggestion
            self._state.columns_by_labware_id[labware_id] = [
                column for column in definition.ordering
            ]
```

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 15]
```diff
@@ -73,6 +81,28 @@ def handle_action(self, action: Action) -> None:
...
 
+        elif isinstance(action, AddPipetteConfigAction):
+            self._state.channels_by_pipette_id[action.pipette_id] = action.channels
+
+    def _set_used_tips(self, pipette_id: str, well_name: str, labware_id: str) -> None:
+        pipette_channels = self._state.channels_by_pipette_id.get(pipette_id)
+        columns = self._state.column_by_labware_id[labware_id]
+        tips_in_labware = self._state.tips_by_labware_id[labware_id]
+
+        if pipette_channels == len(tips_in_labware.keys()):
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/tips.py]: I think the `keys` here is not necessary

```suggestion
        if pipette_channels == len(tips_in_labware):
```

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 16]
```diff
@@ -87,22 +117,40 @@ def __init__(self, state: TipState) -> None:
         """
         self._state = state
 
-    def get_next_tip(
-        self, labware_id: str, use_column: bool, starting_tip_name: Optional[str]
+    def get_next_tip(  # noqa: C901
+        self, labware_id: str, tip_amount: int, starting_tip_name: Optional[str]
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/tips.py]:
```suggestion
        self, labware_id: str, num_tips: int, starting_tip_name: Optional[str]
```

File: api/tests/opentrons/protocol_engine/state/test_tip_state.py [hunk 17]
```diff
@@ -71,77 +84,80 @@ def test_get_next_tip_returns_none(
...
+        tip_amount=input_tip_amount,
         starting_tip_name="B1",
     )
 
-    assert result == "B1"
+    assert result == result_well_name
 
 
+@pytest.mark.parametrize(
+    "input_tip_amount, result_well_name", [(1, "B1"), (8, "A2"), (96, None)]
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_tip_state.py]: I think this test should use two different parameters for tip count:

- Number of channels of the pipette in the `pickUpTip` command
- Number of tips requested for the subsequent `get_next_tip` call

For example, we're going to want to be able to verify an interaction like:

1. A single channel pipette picks up a tip from the middle of a column (e.g. `B1`)
2. `get_next_tip` is called for an 8-channel
3. Verify that the `A1` column is skipped because `B1` is used, returning `A2`

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 18]
```diff
@@ -23,7 +28,9 @@ class TipState:
     """State of all tips."""
 
     tips_by_labware_id: Dict[str, TipRackStateByWellName]
-    column_heads_by_labware_id: Dict[str, Set[str]]
+    # TODO (tz, 12-6-22): should this be a set?
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/tips.py]: No, `Set`s are unordered, which could break stuff

**Review by mcous — COMMENTED** [review]
I think we could use a few more test cases to ensure that `get_next_tip` reacts properly to pickups in the middle of the rack, but this is looking pretty good!

I couldn't come up with any state structures that were meaningfully better than what you've put in place already, so I think the priority here is getting things working with a good test suite in place

File: api/src/opentrons/protocol_engine/commands/load_labware.py (hunk 12 above)
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/commands/load_labware.py]:
```suggestion
        # OpenAPI. The server is allowed to omit it or make it null.
```

File: api/src/opentrons/protocol_engine/state/tips.py (hunk 18 above)
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]:
```suggestion
```

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 19]
```diff
@@ -87,22 +116,50 @@ def __init__(self, state: TipState) -> None:
...
+                            starting_column_index = idx + 1
+                        else:
+                            starting_column_index = idx
+
+            for column in columns[starting_column_index:]:
+                if not any(wells[well] == TipRackWellState.USED for well in column):
+                    return column[0]
+
+        elif num_tips == len(wells.keys()):
+            # not checking against starting_tip_name. for 96c checking that all wells are clean.
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/tips.py]: @Laura-Danielle @mcous for 96 channels I am checking that the whole tip-rack is available without checking against the supplied start tip. Is this the case or should I check if its not the first well return None.

Comment (Laura-Danielle) [reply api/src/opentrons/protocol_engine/state/tips.py]: You could choose to raise an error if the tip supplied is not the first tip of the tiprack since we do not currently support partial tip pick up of the 96 channel.

File: api/src/opentrons/protocol_engine/state/pipettes.py [hunk 20]
```diff
@@ -52,6 +53,14 @@ class CurrentWell:
     well_name: str
 
 
+class StaticPipetteConfig(BaseModel):
+    """Static config for a pipette."""
+
+    model: str
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/state/pipettes.py:59]: nit: if we expect model to be the current load name format, we should type this as `PipetteName` from `shared-data`. For the future, I believe we'll just list the number of channels and the model will actually be of `PipetteModelType` (i.e. `p50` or `p1000`).

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 21]
```diff
@@ -73,6 +80,28 @@ def handle_action(self, action: Action) -> None:
...
+        elif isinstance(action, AddPipetteConfigAction):
+            self._state.channels_by_pipette_id[action.pipette_id] = action.channels
+
+    def _set_used_tips(self, pipette_id: str, well_name: str, labware_id: str) -> None:
+        pipette_channels = self._state.channels_by_pipette_id.get(pipette_id)
+        columns = self._state.column_by_labware_id[labware_id]
+        tips_in_labware = self._state.tips_by_labware_id[labware_id]
+
+        if pipette_channels == len(tips_in_labware):
+            for well_name in tips_in_labware.keys():
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/state/tips.py]: this shadow variable is a bit confusing

File: api/src/opentrons/protocol_engine/state/tips.py [hunk 22]
```diff
@@ -87,22 +116,50 @@ def __init__(self, state: TipState) -> None:
...
-                and tip_state == TipRackWellState.CLEAN
-                and (use_column is False or well_name in column_heads)
-            ):
-                return well_name
+                if seen_start and tip_state == TipRackWellState.CLEAN:
+                    return well_name
 
         return None
+
+    def get_channels(self, pipette_id: str) -> int:
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/state/tips.py]: nit, since this is on the tip state right now could we call this `get_pipette_channels` to avoid any potential confusion there.

**Review by Laura-Danielle — APPROVED** [review]
Had a few nit comments, otherwise looks good to me.

File: api/src/opentrons/protocol_engine/state/pipettes.py (hunk 20 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/pipettes.py:59]: I believe this is the actual model string, rather than what we call the `PipetteName` (it comes from `PipetteDict["model"]`). So `p300_single_v2.1` (model) vs `p300_single_gen2` (name)

**Review by mcous — APPROVED** [review]
🦈
